### ISSUE TITLE
Add a way to handle registed nick in config file

### ISSUE THREAD

**Issue by meain** [osa1/tiny#30]
I am using `auto_cmds` as of now to register my nick.
```yaml
auto_cmds:
      - 'nick meain'
      - 'msg NickServ identify notmypassword'
      - 'join #python'
```

But `join #python` redirects me to `#python-unregistered` as nick is not registered.
The issue is probably because `join #python` is excecuted before nick is approved. Is there a way to deal with this?

**Comment by osa1** [comment]
Hmm.. we could certainly add another section in config for commands that'll only run when nickserv accepts identification .. But that would be too ad-hoc probably. Do you know how/whether other IRC clients do this?

**Comment by meain** [comment]
Irssi does that if we add a config like
```
  Freenode = {
    type = "IRC";
    sasl_mechanism = "PLAIN";
    sasl_username = "meain";
    sasl_password = "asdfasdfasd";
  };
```

right now I am doing something like this in irssi when not using the above method (kind of a hack):
```
 auto_cmd = "nick meain; wait 2000; /msg NickServ IDENTIFY $FREENODE_PASSWORD";
```

**Comment by osa1** [comment]
[SASL](http://ircv3.net/specs/extensions/sasl-3.1.html) method looks easy to implement (assuming it doesn't require TLS). You said "when not using the above method", is this because some servers don't support that? I'm wondering if we'll need a "post-identification" command section in the config anyway.

**Comment by meain** [comment]
I mentioned that just to show a different way.

Also about post auth commands, I think there should be some way to call the join channel commands after the auth part is done.

**Comment by osa1** [comment]
> Also about post auth commands, I think there should be some way to call the join channel commands after the auth part is done.

Right.. As far as I understand, with SASL servers only accept registration (and send the welcome message etc.) after successful login, which is when tiny runs `auto_cmds`. So with SASL I think we'd get that for free.

For the general case I'm not sure what to do.

A `wait` command like you showed works but it's fragile.

We could try to run commands after a nickserv reply but it seems like nickserv in different servers are sending different responses. For example, on freenode it says `You are not identified ...`, on mozilla it says `Password accepted ...`.

**Comment by meain** [comment]
In case of irssi, from what I can tell it does the commands when a response from the irc server comes back ( whether or not the auth was successful ).

So maybe something like doing the auto commands after we receive a response on the auth request we send?

**Comment by osa1** [comment]
Here's an idea:

- We add an optional `auth` field to servers in `.tinyrc.yml`. It looks like this:

    ```yml
    - addr: chat.freenode.net
    - ...
    - auth:
      - service: 'NickServ'
      - msg: 'identify <password>'
    - auto_cmds:
      - 'join #python'
    ```

- When `auth` field is there, we run `auto_cmds` only after we hear back from the "service" mentioned in `auth`'s `service` field. Note that we don't check the reply (message contents etc.), we just run `auto_cmds` after hearing anything from the "service".

- When `auth` is not specified everything works as before.

How does that sound?

In addition, I want to support SASL authentication in the future. With SASL `auth` can take another form:

```
- auth:
  - sasl_username: <sasl username>
  - sasl_password: <sasl password>
```

Because in this method server only accepts the connection on successful login, we can directly run `auto_cmds`.

**Comment by meain** [comment]
@osa1 Sounds good to me

**Comment by osa1** [comment]
So I just implemented this but unfortunately it won't work. The problem is this part:

> When auth field is there, we run auto_cmds only after we hear back from the
> "service" mentioned in auth's service field.

The problem is NickServ sends this message even when I send `/msg NickServ
identify ...` command right after `NICK` and `USER` commands

```
NickServ: This nickname is registered. Please choose a different nickname, or
identify via /msg NickServ identify <password>.
```

So we have to wait until the second message ... but that won't work if NickServ
sends not one but some other number of messages. For example, on
irc.mozilla.org NickServ sends these 4 messages:

```
NickServ: This nickname is registered and protected.  If it is your
NickServ: nick, type /msg NickServ IDENTIFY password.  Otherwise,
NickServ: please choose a different nick.
NickServ: Password accepted - you are now recognized.
```

So counting number of messages coming from the authentication service won't work.

I should probably look at irssi or hexchat source for how they're doing this.

Anyone here have any other ideas?

**Comment by osa1** [comment]
One workaround until we figure this out is to use the new `pass` field. It turns out it's possible to identify in freenode with the `PASS` command. Source: https://superuser.com/questions/374012/irssi-tries-to-join-python-before-sending-the-identify-message

So instead of this

```
servers:
    - addr: chat.freenode.net
      port: 6697
      tls: true
      hostname: omer
      realname: omer
      nicks: [osa1]
      auto_cmds:
          - 'msg NickServ identify password'
          - 'join #python'
```

Do this:

```
servers:
    - addr: chat.freenode.net
      port: 6697
      tls: true
      hostname: omer
      realname: omer
      nicks: [osa1]
      pass: 'password'
      auto_cmds:
          - 'join #python'
```

(nickserv command moved to `pass`)

**Comment by meain** [comment]
@osa1 Sweet, that works.

**Comment by wrycode** [comment]
Awesome. Works on the mozilla chat as well. Since the PASS command is present in both the 1993 and 2000 rfcs, maybe we should put this in the README and default configuration file? The upcoming [ircv3 specifications](https://github.com/ircv3/ircv3-specifications) appear to be backwards compatable with this method too.

**Comment by osa1** [comment]
Added some words about this method of identification in the README and updated the default config to show how to use `pass` field.

Can we close this issue? It seems like most servers support this way of identification so no need add any new features for this.

**Comment by wrycode** [comment]
Yes. If you are still planning to add SASL authentication, it should probably get its own issue.
